### PR TITLE
feat: currency pair validation for defi assets

### PR DESCRIPTION
**Pull request by aljo242**
Closes: BLO-1506

Adds support for formatting assets (base or quote) as:

`ASSET,tokenAddress,chainID`

some extra files touched from running `make format`

This means we now support the following CP formats:

1. `BASE/QUOTE`
2. `BASE/QUOTE,tokenAddress,chainID`
3. `BASE,tokenAddress,chainID/QUOTE`
4. `BASE,tokenAddress,chainID/QUOTE,tokenAddress,chainID`

### CODE REVIEW

File: pkg/types/currency_pair.go [hunk 1]
```diff
@@ -6,8 +6,10 @@ import (
 )
 
 const (
-	ethereum         = "ETHEREUM"
-	MaxCPFieldLength = 128
+	ethereum            = "ETHEREUM"
+	MaxCPFieldLength    = 256
```
Comment (aljo242) [inline pkg/types/currency_pair.go:10]: giving us some headroom but this is kinda arbitrary

File: pkg/types/currency_pair.go [hunk 2]
```diff
@@ -18,9 +20,123 @@ func NewCurrencyPair(base, quote string) CurrencyPair {
...
+		err := ValidateDefiAssetString(cp.Quote)
+		if err != nil {
+			return fmt.Errorf("quote defi asset '%s' is invalid: %w", cp.Quote, err)
+		}
+	}
+
+	return nil
+}
+
+// ValidateLegacyAssetString checks if all fields in the asset string are formatted correctly, i.e.
```
Comment (chenyaoy) [inline pkg/types/currency_pair.go]:
```suggestion
// ValidateLegacyAssetString checks if the asset string is formatted correctly, i.e.
```

File: pkg/types/currency_pair.go [hunk 3]
```diff
@@ -18,9 +20,123 @@ func NewCurrencyPair(base, quote string) CurrencyPair {
...
+// - fields are fully uppercase
+// - fields
```
Comment (chenyaoy) [inline pkg/types/currency_pair.go]:
```suggestion
// - asset string does not contain the `fieldSeparator`
```

File: pkg/types/currency_pair.go [hunk 4]
```diff
@@ -18,9 +20,123 @@ func NewCurrencyPair(base, quote string) CurrencyPair {
...
+		if err != nil {
+			return fmt.Errorf("quote defi asset '%s' is invalid: %w", cp.Quote, err)
+		}
+	}
+
+	return nil
+}
+
+// ValidateLegacyAssetString checks if all fields in the asset string are formatted correctly, i.e.
+// - fields are fully uppercase
```
Comment (chenyaoy) [inline pkg/types/currency_pair.go]:
```suggestion
// - asset string is fully uppercase
```

**Review by chenyaoy — COMMENTED** [review]
lgtm

File: pkg/types/currency_pair.go [hunk 5]
```diff
@@ -18,9 +20,123 @@ func NewCurrencyPair(base, quote string) CurrencyPair {
...
+// - fields
+//
+// NOTE: this function assumes that coreValidate() has already been run.
+func ValidateLegacyAssetString(asset string) error {
+	// check formatting of asset
+	if strings.ToUpper(asset) != asset {
+		return fmt.Errorf("incorrectly formatted asset string, expected: %s got: %s", strings.ToUpper(asset), asset)
+	}
+
+	split := strings.Split(asset, fieldSeparator)
```
Comment (vincentwschau) [inline pkg/types/currency_pair.go]: nit: Why not just use `IsLegacyAssetString` again to verify this?

File: pkg/types/currency_pair.go [hunk 6]
```diff
@@ -18,9 +20,123 @@ func NewCurrencyPair(base, quote string) CurrencyPair {
...
+	// first element is a ticker, so we require it to pass legacy asset validation:
+	if err := ValidateLegacyAssetString(split[0]); err != nil {
+		return fmt.Errorf("asset field '%s' is invalid: %w", split[0], err)
+	}
+
+	return nil
+}
+
+// splitDefiAssetString splits a defi asset by the fieldSeparator and checks that it is the proper length.
+func splitDefiAssetString(asset string) ([]string, error) {
```
Comment (vincentwschau) [inline pkg/types/currency_pair.go]: nit: A public utility function that splits the string and returns some named string return values would be useful for logic importing this struct.

**Review by chenyaoy — APPROVED** [review]

**Review by vincentwschau — APPROVED** [review]
LGTM

File: abci/ve/vote_extension_test.go [hunk 7]
```diff
@@ -677,7 +677,7 @@ func (s *VoteExtensionTestSuite) TestExtendVoteStatus() {
 
 	s.Run("test price transformation failures", func() {
 		mockMetrics := metricsmocks.NewMetrics(s.T())
-		transformationError := fmt.Errorf("incorrectly formatted CurrencyPair: BTCETH")
+		transformationError := fmt.Errorf("incorrectly formatted CurrencyPair: \"BTCETH\"")
```
Comment (technicallyty) [inline abci/ve/vote_extension_test.go:680]: nit: i believe you can use the back tick \` and then you dont have to escape quotes

File: pkg/types/currency_pair.go [hunk 8]
```diff
@@ -18,9 +20,123 @@ func NewCurrencyPair(base, quote string) CurrencyPair {
...
+	// check quote asset
+	if IsLegacyAssetString(cp.Quote) {
+		err := ValidateLegacyAssetString(cp.Quote)
+		if err != nil {
+			return fmt.Errorf("quote asset '%s' is invalid: %w", cp.Quote, err)
+		}
+	} else {
+		err := ValidateDefiAssetString(cp.Quote)
+		if err != nil {
+			return fmt.Errorf("quote defi asset '%s' is invalid: %w", cp.Quote, err)
```
Comment (technicallyty) [inline pkg/types/currency_pair.go]:
```suggestion
			return fmt.Errorf("base asset %q is invalid: %w", cp.Base, err)
		}
	} else {
		err := ValidateDefiAssetString(cp.Base)
		if err != nil {
			return fmt.Errorf("base defi asset %q is invalid: %w", cp.Base, err)
		}
	}

	// check quote asset
	if IsLegacyAssetString(cp.Quote) {
		err := ValidateLegacyAssetString(cp.Quote)
		if err != nil {
			return fmt.Errorf("quote asset %q is invalid: %w", cp.Quote, err)
		}
	} else {
		err := ValidateDefiAssetString(cp.Quote)
		if err != nil {
			return fmt.Errorf("quote defi asset %q is invalid: %w", cp.Quote, err)
```
`%q` does this for you

File: pkg/types/currency_pair.go [hunk 9]
```diff
@@ -18,9 +20,123 @@ func NewCurrencyPair(base, quote string) CurrencyPair {
...
+// NOTE: this function assumes that coreValidate() has already been run.
+func ValidateDefiAssetString(asset string) error {
+	token, _, _, err := SplitDefiAssetString(asset)
+	if err != nil {
+		return err
+	}
+
+	// first element is a ticker, so we require it to pass legacy asset validation:
+	if err := ValidateLegacyAssetString(token); err != nil {
+		return fmt.Errorf("token field '%s' is invalid: %w", token, err)
```
Comment (technicallyty) [inline pkg/types/currency_pair.go]: %q

File: pkg/types/currency_pair.go [hunk 10]
```diff
@@ -18,9 +20,123 @@ func NewCurrencyPair(base, quote string) CurrencyPair {
...
+		return fmt.Errorf("quote currency pair cannot be empty")
+	}
+
+	if len(cp.Base) > MaxCPFieldLength {
+		return fmt.Errorf("base currency pair has wrong length")
+	}
+
+	if len(cp.Quote) > MaxCPFieldLength {
+		return fmt.Errorf("quote currency pair has wrong length")
+	}
```
Comment (technicallyty) [inline pkg/types/currency_pair.go]: may want to specify what the max is in these errors

File: pkg/types/currency_pair.go [hunk 11]
```diff
@@ -59,21 +175,50 @@ func CurrencyPairString(base, quote string) string {
...
+		s = strings.ToUpper(s)
+	} else {
+		token, address, chainID, err := SplitDefiAssetString(s)
+		if err != nil {
+			return "", fmt.Errorf("incorrectly formatted asset: %q: %w", s, err)
+		}
+
+		token = strings.ToUpper(token)
+		s = strings.Join([]string{token, address, chainID}, fieldSeparator)
+	}
```
Comment (technicallyty) [inline pkg/types/currency_pair.go:217]: we only upper the token?

Comment (aljo242) [reply pkg/types/currency_pair.go:217]: yes, wnat to preserve any formatting that might make addresses / chainIDs unique

File: pkg/types/currency_pair.go [hunk 12]
```diff
@@ -18,19 +20,135 @@ func NewCurrencyPair(base, quote string) CurrencyPair {
...
+
+	return nil
+}
+
+// ValidateLegacyAssetString checks if the asset string is formatted correctly, i.e.
+// - asset string is fully uppercase
+// - asset string does not contain the `fieldSeparator`
+//
+// NOTE: this function assumes that coreValidate() has already been run.
+func ValidateLegacyAssetString(asset string) error {
```
Comment (technicallyty) [inline pkg/types/currency_pair.go:95]: if someone were to use this function outside of this package, then `coreValidate` won't be visible to them. what do they need to run to trigger a coreValidate?

**Review by technicallyty — APPROVED** [review]
LGMT, just one nit with the comment on `ValidateLegacyAssetString` and `ValidateDefiAssetString`.

**Review by davidterpay — APPROVED** [review]

**Pull request merged**